File: compilers/java/src/test/java/runar/compiler/passes/UnknownAnfKindTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import runar.compiler.ir.UnknownAnfKindError;
import runar.compiler.ir.anf.AnfBinding;
import runar.compiler.ir.anf.AnfMethod;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.anf.AnfValue;
import runar.compiler.ir.anf.BoolConst;
import runar.compiler.ir.anf.Call;
import runar.compiler.ir.anf.If;
import runar.compiler.ir.anf.LoadConst;
import runar.compiler.ir.anf.SyntheticAnfValueForTests;

/**
 * Regression test for F-003 (Java tier): every ANF-kind dispatch in the
 * Java compiler must throw {@link UnknownAnfKindError} when it
 * encounters a kind it doesn't recognize, instead of silently returning
 * an empty / no-op result.
 *
 * <p>Each test drives one dispatch site with a {@link SyntheticAnfValueForTests}
 * instance whose {@code kind()} does not appear in any production
 * dispatcher, then asserts the resulting throw is the typed error and
 * carries the synthetic kind name and a location string identifying the
 * dispatcher.
 *
 * <p>Mirrors {@code packages/runar-compiler/src/__tests__/unknown-anf-kind.test.ts}.
 *
 * <p>If a new {@link AnfValue} variant is added in the future, the
 * dispatch sites below must be updated; this test guards against
 * silently shipping an unhandled variant.
 */
class UnknownAnfKindTest {

    private static final String SYNTHETIC_KIND = "synthetic_test_kind_for_regression_only";

    private static AnfValue syntheticValue() {
        return SyntheticAnfValueForTests.of(SYNTHETIC_KIND);
    }

    private static AnfProgram makeProgram(List<AnfBinding> body) {
        AnfMethod method = new AnfMethod("m", List.of(), body, true);
        return new AnfProgram("Test", List.of(), List.of(method));
    }

    @Test
    void errorMessageReferencesDeveloperRecipe() {
        UnknownAnfKindError err = new UnknownAnfKindError(SYNTHETIC_KIND, "unit-test.location");
        assertTrue(err.getMessage().contains(SYNTHETIC_KIND),
            "message must contain the synthetic kind: " + err.getMessage());
        assertTrue(err.getMessage().contains("unit-test.location"),
            "message must contain the location: " + err.getMessage());
        assertTrue(err.getMessage().contains("Adding a New ANF Value Kind"),
            "message must reference the developer recipe: " + err.getMessage());
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertEquals("unit-test.location", err.getLocation());
    }

    @Test
    void throwsFromConstantFoldFoldValue() {
        AnfProgram program = makeProgram(List.of(
            new AnfBinding("t0", syntheticValue(), null)
        ));

        UnknownAnfKindError err = assertThrows(UnknownAnfKindError.class,
            () -> ConstantFold.run(program));
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertEquals("constant-fold.foldValue", err.getLocation());
    }

    @Test
    void throwsFromAnfOptimizeDispatchers() {
        // AnfOptimize.run is gated on the presence of an EC call (`ecAdd`,
        // `ecMul`, ...) — bodies with no EC primitives return unchanged and
        // never reach the dispatchers, by design. We seed the body with a
        // benign ecAdd call so the gate trips, then place the synthetic kind
        // in a sibling binding so renameInValue / collectRefs / hasSideEffect
        // walk over it and throw.
        AnfProgram program = makeProgram(List.of(
            new AnfBinding("ec0", new Call("ecAdd", List.of("a", "b")), null),
            new AnfBinding("t0", syntheticValue(), null)
        ));

        UnknownAnfKindError err = assertThrows(UnknownAnfKindError.class,
            () -> AnfOptimize.run(program));
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertTrue(
            err.getLocation().equals("anf-optimize.renameInValue")
                || err.getLocation().equals("anf-optimize.collectRefs")
                || err.getLocation().equals("anf-optimize.hasSideEffect"),
            "expected an anf-optimize dispatcher location, got " + err.getLocation());
    }

    @Test
    void throwsFromAnfLowerRemapValueRefs() {
        // remapValueRefs is package-private so we can drive it directly,
        // matching the TS test that calls remapValueRefs(synthetic, {}).
        UnknownAnfKindError err = assertThrows(UnknownAnfKindError.class,
            () -> AnfLower.remapValueRefs(syntheticValue(), new HashMap<>()));
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertEquals("anf-lower.remapValueRefs", err.getLocation());
    }

    @Test
    void throwsFromStackLower() {
        // collectRefs runs first inside StackLower.run (computeLastUses);
        // lowerBinding is the fallback path. Both are acceptable, matching
        // the TS test.
        AnfProgram program = makeProgram(List.of(
            new AnfBinding("t0", syntheticValue(), null)
        ));

        UnknownAnfKindError err = assertThrows(UnknownAnfKindError.class,
            () -> StackLower.run(program));
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertTrue(
            err.getLocation().equals("stack-lower.collectRefs")
                || err.getLocation().equals("stack-lower.lowerBinding"),
            "expected a stack-lower dispatcher location, got " + err.getLocation());
    }

    @Test
    void throwsFromAnfLoaderOnUnknownKindJson() {
        // The loader rejects unknown kinds at JSON decode time. Wrap the
        // synthetic kind inside a minimal well-formed AnfProgram JSON so
        // the parser walks from program -> method -> binding -> value and
        // hits the unknown-kind branch in toValue.
        String json = "{"
            + "\"contractName\":\"T\","
            + "\"properties\":[],"
            + "\"methods\":[{"
            +   "\"name\":\"m\",\"isPublic\":true,\"params\":[],"
            +   "\"body\":[{\"name\":\"t0\",\"value\":{\"kind\":\"" + SYNTHETIC_KIND + "\"}}]"
            + "}]"
            + "}";

        UnknownAnfKindError err = assertThrows(UnknownAnfKindError.class,
            () -> AnfLoader.parse(json));
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertEquals("anf-loader.parseValue", err.getLocation());
    }

    @Test
    void stackLowerCollectRefsExhaustsKnownKinds() {
        // Defensive: also confirm collectRefs is reached when the synthetic
        // value is nested under an `if`. The outer `if` is a known kind, so
        // the dispatcher must recurse into the then-branch and throw there.
        AnfProgram program = makeProgram(List.of(
            new AnfBinding("c", new LoadConst(new BoolConst(true)), null),
            new AnfBinding(
                "t0",
                new If(
                    "c",
                    List.of(new AnfBinding("tn", syntheticValue(), null)),
                    List.of()
                ),
                null
            )
        ));

        UnknownAnfKindError err = assertThrows(UnknownAnfKindError.class,
            () -> StackLower.run(program));
        assertEquals(SYNTHETIC_KIND, err.getKind());
        assertNotNull(err.getLocation());
        assertTrue(err.getLocation().startsWith("stack-lower."),
            "expected a stack-lower location, got " + err.getLocation());
    }
}

File: compilers/go/frontend/unknown_anf_kind_test.go
// Package frontend — regression test for F-003: every ANF-kind dispatch
// in the Go ANF-lowering / constant-fold / DCE passes must reject an
// unknown kind with a typed *ir.UnknownANFKindError, instead of
// silently returning the value unchanged / dropping refs / defaulting
// to "no side effect".
//
// Each test drives one dispatch site with a synthetic ANFValue whose
// `Kind` does not appear in the ANFValue schema, then asserts the
// resulting panic carries the typed error with the synthetic kind name.
package frontend

import (
	"strings"
	"testing"

	"github.com/icellan/runar/compilers/go/ir"
)

const syntheticFrontendKind = "synthetic_test_kind_for_regression_only"

func syntheticFrontendValue() ir.ANFValue {
	return ir.ANFValue{Kind: syntheticFrontendKind}
}

// recoverFrontendUnknownKind runs fn and recovers a panic. It fails the
// test if the panic value was not *ir.UnknownANFKindError.
func recoverFrontendUnknownKind(t *testing.T, fn func()) *ir.UnknownANFKindError {
	t.Helper()
	var caught *ir.UnknownANFKindError
	func() {
		defer func() {
			if r := recover(); r != nil {
				if e, ok := r.(*ir.UnknownANFKindError); ok {
					caught = e
					return
				}
				t.Fatalf("expected *ir.UnknownANFKindError panic, got %T: %v", r, r)
			}
		}()
		fn()
	}()
	return caught
}

func TestUnknownANFKind_ConstantFold_FoldValue(t *testing.T) {
	v := syntheticFrontendValue()
	env := newConstEnv()
	err := recoverFrontendUnknownKind(t, func() {
		_ = foldValue(&v, env)
	})
	if err == nil {
		t.Fatal("expected foldValue panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	if err.Location != "constant-fold.foldValue" {
		t.Errorf("expected location constant-fold.foldValue, got %q", err.Location)
	}
}

func TestUnknownANFKind_AnfLower_RemapValueRefs(t *testing.T) {
	v := syntheticFrontendValue()
	err := recoverFrontendUnknownKind(t, func() {
		_ = remapValueRefs(v, map[string]string{})
	})
	if err == nil {
		t.Fatal("expected remapValueRefs panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	if err.Location != "anf-lower.remapValueRefs" {
		t.Errorf("expected location anf-lower.remapValueRefs, got %q", err.Location)
	}
}

func TestUnknownANFKind_AnfLower_IsSideEffectFree(t *testing.T) {
	v := syntheticFrontendValue()
	err := recoverFrontendUnknownKind(t, func() {
		_ = isSideEffectFree(&v)
	})
	if err == nil {
		t.Fatal("expected isSideEffectFree panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	if err.Location != "anf-lower.isSideEffectFree" {
		t.Errorf("expected location anf-lower.isSideEffectFree, got %q", err.Location)
	}
}

func TestUnknownANFKind_AnfOptimize_CollectValueRefs(t *testing.T) {
	v := syntheticFrontendValue()
	refs := map[string]bool{}
	err := recoverFrontendUnknownKind(t, func() {
		collectValueRefs(&v, refs)
	})
	if err == nil {
		t.Fatal("expected collectValueRefs panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	if err.Location != "anf-optimize.collectValueRefs" {
		t.Errorf("expected location anf-optimize.collectValueRefs, got %q", err.Location)
	}
}

func TestUnknownANFKind_AnfOptimize_HasSideEffect(t *testing.T) {
	v := syntheticFrontendValue()
	err := recoverFrontendUnknownKind(t, func() {
		_ = hasSideEffect(&v)
	})
	if err == nil {
		t.Fatal("expected hasSideEffect panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	if err.Location != "anf-optimize.hasSideEffect" {
		t.Errorf("expected location anf-optimize.hasSideEffect, got %q", err.Location)
	}
}

func TestUnknownANFKind_AnfOptimize_EliminateDeadBindings(t *testing.T) {
	// Drive DCE end-to-end with a synthetic-kind binding in the body.
	// The first dispatch in eliminateDeadBindings is collectAllRefs ->
	// collectValueRefs, so we expect that location to fire first.
	method := &ir.ANFMethod{
		Name:     "m",
		Params:   nil,
		IsPublic: true,
		Body: []ir.ANFBinding{
			{Name: "t0", Value: syntheticFrontendValue()},
		},
	}
	err := recoverFrontendUnknownKind(t, func() {
		eliminateDeadBindings(method)
	})
	if err == nil {
		t.Fatal("expected eliminateDeadBindings panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	// Either collectValueRefs or hasSideEffect can be the first to fire.
	switch err.Location {
	case "anf-optimize.collectValueRefs", "anf-optimize.hasSideEffect":
	default:
		t.Errorf("expected anf-optimize.{collectValueRefs,hasSideEffect}, got %q", err.Location)
	}
}

func TestUnknownANFKind_FoldConstants_TopLevel(t *testing.T) {
	// Drive constant-fold end-to-end. foldValue is called per binding,
	// so the synthetic kind should reach the default branch.
	program := &ir.ANFProgram{
		ContractName: "T",
		Methods: []ir.ANFMethod{
			{
				Name:     "m",
				Params:   nil,
				IsPublic: true,
				Body: []ir.ANFBinding{
					{Name: "t0", Value: syntheticFrontendValue()},
				},
			},
		},
	}
	err := recoverFrontendUnknownKind(t, func() {
		_ = FoldConstants(program)
	})
	if err == nil {
		t.Fatal("expected FoldConstants panic for unknown kind")
	}
	if err.Kind != syntheticFrontendKind {
		t.Errorf("expected kind %q, got %q", syntheticFrontendKind, err.Kind)
	}
	if !strings.Contains(err.Location, "constant-fold") {
		t.Errorf("expected constant-fold location, got %q", err.Location)
	}
}
